<launch>
  <arg name="model" default="$(env TURTLEBOT3_MODEL)" doc="model type [burger, waffle, waffle_pi]"/>
  <arg name="first_tb3"  default="tb3_0"/>
  <arg name="second_tb3" default="tb3_1"/>
  <arg name="third_tb3"  default="tb3_2"/>
  <arg name="dim_state"  default="6"/>
  <arg name="marker_length"  default="0.22884"/>
  <arg name="aruco_cal_file"  default="calibrationSave_2.p"/>

  <arg name="first_tb3_x_pos" default="-2.0"/>
  <arg name="first_tb3_y_pos" default="-1.0"/>
  <arg name="first_tb3_z_pos" default=" 0.0"/>
  <arg name="first_tb3_yaw"   default=" 0.524"/>
  <arg name="first_tb3_id"    default=" 2001"/>

  <arg name="second_tb3_x_pos" default=" 2.0"/>
  <arg name="second_tb3_y_pos" default="-1.0"/>
  <arg name="second_tb3_z_pos" default=" 0.0"/>
  <arg name="second_tb3_yaw"   default=" 2.618"/>
  <arg name="second_tb3_id"    default=" 2002"/>

  <arg name="third_tb3_x_pos" default=" 0.0"/>
  <arg name="third_tb3_y_pos" default=" 2.0"/>
  <arg name="third_tb3_z_pos" default=" 0.0"/>
  <arg name="third_tb3_yaw"   default=" 4.712"/>
  <arg name="third_tb3_id"    default=" 2003"/>

  <!--<include file="$(find dse_simulation)/launch/my_world.launch">
  </include>-->

  <node pkg="dse_simulation" type="gazebo_sim_visualize_pose_world.py" name="dse_gazebo_object_visualization_node" output="screen">
    <param name="n_params" value="3" />
    <param name="objects" value="[aruco_marker_0, aruco_marker_2, aruco_marker_3]" />
  </node>

  <group ns = "$(arg first_tb3)">
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(find turtlebot3_description)/urdf/turtlebot3_$(arg model).urdf.xacro" />

    <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" output="screen">
      <param name="publish_frequency" type="double" value="50.0" />
      <param name="tf_prefix" value="$(arg first_tb3)" />
    </node>

    <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-urdf -model $(arg first_tb3) -x $(arg first_tb3_x_pos) -y $(arg first_tb3_y_pos) -z $(arg first_tb3_z_pos) -Y $(arg first_tb3_yaw) -param robot_description" />

    <node pkg="dse_simulation" type="aruco_pose_estimation" name="aruco_pose_estimation_node" output="screen">
      <param name="prefix" value="$(arg first_tb3)" />
      <param name="marker_length" type="double" value="$(arg marker_length)" />
      <param name="calibration_file" value="$(arg aruco_cal_file)" />
    </node>
    <node pkg="dse_simulation" type="information_filter" name="information_filter_node" output="screen">
      <param name="prefix" value="$(arg first_tb3)" />
      <param name="id" type="int" value="$(arg first_tb3_id)" />
      <param name="dim_state" type="int" value="$(arg dim_state)" />
    </node>
    <node pkg="dse_simulation" type="direct_estimator" name="direct_estimator_node" output="screen">
      <param name="prefix" value="$(arg first_tb3)" />
    </node>
    <node pkg="dse_simulation" type="rqt_plot_output.py" name="dse_plotting_node" output="screen">
      <param name="prefix" value="$(arg first_tb3)" />
      <param name="id" type="int" value="$(arg first_tb3_id)" />
      <param name="dim_state" type="int" value="$(arg dim_state)" />
    </node>
    <node pkg="dse_simulation" type="gazebo_sim_visualize_pose_base.py" name="dse_gazebo_visualization_node" output="screen">
      <param name="prefix" value="$(arg first_tb3)" />
      <param name="id" type="int" value="$(arg first_tb3_id)" />
      <param name="dim_state" type="int" value="$(arg dim_state)" />
    </node>
    <node pkg="dse_simulation" type="world_tf_broadcaster.py" name="world_tf_broadcaster" output="screen">
      <param name="gazebo_name" value="$(arg first_tb3)" />
      <param name="tf_name" value="$(arg first_tb3)/odom" />
    </node>
  </group>

  <group ns = "$(arg second_tb3)">
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(find turtlebot3_description)/urdf/turtlebot3_$(arg model).urdf.xacro" />

    <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" output="screen">
      <param name="publish_frequency" type="double" value="50.0" />
      <param name="tf_prefix" value="$(arg second_tb3)" />
    </node>

    <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-urdf -model $(arg second_tb3) -x $(arg second_tb3_x_pos) -y $(arg second_tb3_y_pos) -z $(arg second_tb3_z_pos) -Y $(arg second_tb3_yaw) -param robot_description" />

    <node pkg="dse_simulation" type="aruco_pose_estimation" name="aruco_pose_estimation_node" output="screen">
      <param name="prefix" value="$(arg second_tb3)" />
      <param name="marker_length" type="double" value="$(arg marker_length)" />
      <param name="calibration_file" value="$(arg aruco_cal_file)" />
    </node>
    <node pkg="dse_simulation" type="information_filter" name="information_filter_node" output="screen">
      <param name="prefix" value="$(arg second_tb3)" />
      <param name="id" type="int" value="$(arg second_tb3_id)" />
      <param name="dim_state" type="int" value="$(arg dim_state)" />
    </node>
    <node pkg="dse_simulation" type="direct_estimator" name="direct_estimator_node" output="screen">
      <param name="prefix" value="$(arg second_tb3)" />
    </node>
    <node pkg="dse_simulation" type="rqt_plot_output.py" name="dse_plotting_node" output="screen">
      <param name="prefix" value="$(arg second_tb3)" />
      <param name="id" type="int" value="$(arg second_tb3_id)" />
      <param name="dim_state" type="int" value="$(arg dim_state)" />
    </node>
    <node pkg="dse_simulation" type="gazebo_sim_visualize_pose_base.py" name="dse_gazebo_visualization_node" output="screen">
      <param name="prefix" value="$(arg second_tb3)" />
      <param name="id" type="int" value="$(arg second_tb3_id)" />
      <param name="dim_state" type="int" value="$(arg dim_state)" />
    </node>
    <node pkg="dse_simulation" type="world_tf_broadcaster.py" name="world_tf_broadcaster" output="screen">
      <param name="gazebo_name" value="$(arg second_tb3)" />
      <param name="tf_name" value="$(arg second_tb3)/odom" />
    </node>
  </group>

  <group ns = "$(arg third_tb3)">
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(find turtlebot3_description)/urdf/turtlebot3_$(arg model).urdf.xacro" />

    <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" output="screen">
      <param name="publish_frequency" type="double" value="50.0" />
      <param name="tf_prefix" value="$(arg third_tb3)" />
    </node>

    <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-urdf -model $(arg third_tb3) -x $(arg third_tb3_x_pos) -y $(arg third_tb3_y_pos) -z $(arg third_tb3_z_pos) -Y $(arg third_tb3_yaw) -param robot_description" />

    <node pkg="dse_simulation" type="aruco_pose_estimation" name="aruco_pose_estimation_node" output="screen">
      <param name="prefix" value="$(arg third_tb3)" />
      <param name="marker_length" type="double" value="$(arg marker_length)" />
      <param name="calibration_file" value="$(arg aruco_cal_file)" />
    </node>
    <node pkg="dse_simulation" type="information_filter" name="information_filter_node" output="screen">
      <param name="prefix" value="$(arg third_tb3)" />
      <param name="id" type="int" value="$(arg third_tb3_id)" />
      <param name="dim_state" type="int" value="$(arg dim_state)" />
    </node>
    <node pkg="dse_simulation" type="direct_estimator" name="direct_estimator_node" output="screen">
      <param name="prefix" value="$(arg third_tb3)" />
    </node>
    <node pkg="dse_simulation" type="rqt_plot_output.py" name="dse_plotting_node" output="screen">
      <param name="prefix" value="$(arg third_tb3)" />
      <param name="id" type="int" value="$(arg third_tb3_id)" />
      <param name="dim_state" type="int" value="$(arg dim_state)" />
    </node>
    <node pkg="dse_simulation" type="gazebo_sim_visualize_pose_base.py" name="dse_gazebo_visualization_node" output="screen">
      <param name="prefix" value="$(arg third_tb3)" />
      <param name="id" type="int" value="$(arg third_tb3_id)" />
      <param name="dim_state" type="int" value="$(arg dim_state)" />
    </node>
    <node pkg="dse_simulation" type="world_tf_broadcaster.py" name="world_tf_broadcaster" output="screen">
      <param name="gazebo_name" value="$(arg third_tb3)" />
      <param name="tf_name" value="$(arg third_tb3)/odom" />
    </node>
  </group>
</launch>
